bytecomp/translmod.ml

```diff
@@ -342,9 +353,82 @@ let transl_class_bindings cl_list =
        (id, transl_class ids id meths cl vf))
      cl_list)
 
+(* Compile one or more functors, merging curried functors to produce
+   multi-argument functors.  Any [@inline] attribute on a functor that is
+   merged must be consistent with any other [@inline] attribute(s) on the
+   functor(s) being merged with.  Such an attribute will be placed on the
+   resulting merged functor. *)
+
+let merge_inline_attributes attr1 attr2 loc =
+  match Lambda.merge_inline_attributes attr1 attr2 with
+  | Some attr -> attr
+  | None -> raise (Error (loc, Conflicting_inline_attributes))
+
+let merge_functors mexp coercion root_path =
+  let rec merge mexp coercion path acc inline_attribute =
+    let finished () = acc, mexp, path, coercion, inline_attribute in
+    match mexp.mod_type with
+    | Mty_alias _ -> finished ()
+    | _ ->
+      match mexp.mod_desc with
+      | Tmod_functor (param, _, _, body) ->
+        let inline_attribute' =
+          Translattribute.get_inline_attribute mexp.mod_attributes
+        in
+        let arg_coercion, res_coercion =
+          match coercion with
+          | Tcoerce_none -> None, Tcoerce_none
+          | Tcoerce_functor (arg_coercion, res_coercion) ->
+            Some arg_coercion, res_coercion
+          | _ -> fatal_error "Translmod.merge_functors: bad coercion"
+        in
+        let loc = mexp.mod_loc in
+        let path = functor_path path param in
+        let inline_attribute =
+          merge_inline_attributes inline_attribute inline_attribute' loc
+        in
+        merge body res_coercion path ((param, loc, arg_coercion) :: acc)
+          inline_attribute
+      | _ -> finished ()
+  in
+  merge mexp coercion root_path [] Default_inline
+
+let rec compile_functor mexp coercion root_path loc =
+  let functor_params_rev, body, body_path, res_coercion, inline_attribute =
+    merge_functors mexp coercion root_path
+  in
+  assert (List.length functor_params_rev >= 1);  (* cf. [transl_module] *)
+  let params, body =
+    List.fold_left (fun (params, body) (param, loc, arg_coercion) ->
+        let param' = Ident.rename param in
+        let arg =
+          match arg_coercion with
+          | None -> Lvar param'
+          | Some arg_coercion ->
+            apply_coercion loc Alias arg_coercion (Lvar param')
```

Comment:
If you used `Coerce_none` instead of `None` then this match can be removed since the second case will just evaluate to the first.